### PR TITLE
Add support for picking RD_LOCATION for BATS

### PR DESCRIPTION
**Pull request by jandubois**
Choose from system/user/dist/npm or pick first available. dist and npm only work when running from inside a git clone. npm is not yet supported on Windows.

Also move "load ../helpers/load" from setup() to top level, so we can abort on fatal errors.

Predefine teardown_file() with `rdctl shutdown` so it doesn't have to be repeated in every test file.

Partially implements #4542 but more work is needed.

### Additional comments

`./bats-core/bin/bats tests/*` passed successfully on macOS Intel for both "moby" and "containerd": `158 tests, 0 failures, 5 skipped`.

You can tell BATS to use the package in the `dist` directory even when RD is also installed globally, or per user:

```
RD_LOCATION=dist ./bats-core/bin/bats tests/*
```

I'm not really happy with the `RD_LOCATION` name, but couldn't think of anything better. I briefly considered `RD_INSTALL`, but that seems even worse.

Using `RD_LOCATION=npm` has been tested with the `registry/creds.bats` test, which does run a second factory-reset in the middle:

```console
$ RD_LOCATION=npm ./bats-core/bin/bats tests/registry/creds.bats
creds.bats
[...]

19 tests, 0 failures
```

The factory reset logic should be made more robust (e.g. I've seen it once that limactl/qemu were not killed), and needs to be implemented for Windows. This should happen in a separate PR.

The `apify_arg()` function should be auto-generated from the schema in `command-api.yaml`. This too should happen in a separate PR.

### CODE REVIEW

File: bats/tests/helpers/defaults.bash
Comment (mook-as) [inline bats/tests/helpers/defaults.bash:2]:
```suggestion
    local var="$1"
```

File: bats/tests/helpers/load.bash
Comment (mook-as) [inline bats/tests/helpers/load.bash:13]: `bats-support` has a `fail`; would that be useful?

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:1]: Rather than a comment, we can just assert that they _are_ set (this avoids issues with incorrect load orders).

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:14]: These should be quoted.

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:23]: This would be (slightly) clearer to say `elif [ -f "$dist" ]; then … elif [ -d "$PATH_REPO_ROOT/pkg/rancher-desktop" ]; then…`
If we're not inside a repo checkout, `$dist` wouldn't exist.

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:29]: It might be good to print out the locations we checked.

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:53]: is `DIST_ARCH` used anywhere else? (Can it be `local`?)

File: bats/tests/helpers/vm.bash
Comment (mook-as) [inline bats/tests/helpers/vm.bash:35]: We should consider making `npm run dev` take the same arguments as `rdctl start` instead. That's better for our sanity. :)

File: bats/tests/helpers/defaults.bash
Comment (jandubois) [reply bats/tests/helpers/defaults.bash:2]: Why? It only makes a difference in places where the shell would otherwise perform word-splitting, but that is not the case for variable assignments:

```console
$ bar=$(echo one two)
$ foo=$bar printenv foo
one two
$ foo() { local var=$1; echo "$var"; }
$ foo one two
one
$ foo "one two"
one two
```

Now, we can make the argument that we should always quote shell variables, just so the reader doesn't have to know/understand when the shell will split words, and when it doesn't.

I'm fine with that, except that we just argued in the last retro that we don't want to call out style issues in code review for code that passes the linter and auto-formatter, and in this case both `shellcheck` and `shfmt` are fine with not quoting the variable.

I'm going to add the quotes, but leave it to you to figure out how to teach this rule to our tools, so that we can apply this consistently. Because a simplistic search finds at least another 19 lines (in addition to the 5 from this PR) where we don't quote variable assignments:

```
$ grep -r '=\$' tests | grep -v '"' | wc -l
      24
```

File: bats/tests/helpers/load.bash
Comment (jandubois) [reply bats/tests/helpers/load.bash:13]: I don't think so, as I called my version `fail` initially and then spend some time to figure out why it wouldn't work. If the bat-support version had worked, then I probably wouldn't have noticed. 😄 But I can take another look...

File: bats/tests/helpers/paths.bash
Comment (jandubois) [reply bats/tests/helpers/paths.bash:1]: I'm not worried about load order because `load.bash` is the function that sources everything else. It is just that I expect **all** `PATH_*` variables to be defined in `paths.bash`, so I wanted to document the 2 exceptions. I can remove the comment, if you think it is not useful, but I don't see any value in adding an assertion.

File: bats/tests/helpers/paths.bash
Comment (jandubois) [reply bats/tests/helpers/paths.bash:23]: I don't understand what you are saying. Can you use the "suggestion" feature to show your proposed replacement?

> If we're not inside a repo checkout, $dist wouldn't exist.

Yes, and that's why I check first that we are inside a repo checkout before testing `$dist`. So we are in agreement on this part, but why would you then change the code order?

File: bats/tests/helpers/paths.bash
Comment (jandubois) [reply bats/tests/helpers/paths.bash:53]: It is not used anywhere else. It can't be `local` because we are not inside a function and `local` doesn't limit visibility to an `if` clause.

File: bats/tests/helpers/vm.bash
Comment (jandubois) [reply bats/tests/helpers/vm.bash:35]: That works for me too, but should be a separate PR.

File: bats/tests/helpers/defaults.bash
Comment (jandubois) [reply bats/tests/helpers/defaults.bash]: I've made the changes, also to all other locations that didn't follow the "always quote substitutions" rule: https://github.com/rancher-sandbox/rancher-desktop/commit/40c2deac8053f7c1e9d21cb7aa14dcbcb1828dd1

Not really sure if this is a net gain when I look at things like this:

```diff
-PATH_BATS_HELPERS=$(absolute_path "$(dirname "${BASH_SOURCE[0]}")")
-PATH_BATS_ROOT=$(absolute_path "$PATH_BATS_HELPERS/../..")
+PATH_BATS_HELPERS="$(absolute_path "$(dirname "${BASH_SOURCE[0]}")")"
+PATH_BATS_ROOT="$(absolute_path "$PATH_BATS_HELPERS/../..")"
```

Please don't merge as-is; I have used separate commits for further review, but want to squash before merging.

File: bats/tests/helpers/load.bash
Comment (jandubois) [reply bats/tests/helpers/load.bash:13]: I checked again, and the problem is that the bats `fail` function writes to stderr, which is discarded unless a test failure is reported; we need to write to `&3` to print something that is not captured again by bats.

Comment (mook-as) [reply bats/tests/helpers/load.bash:13]: Ah, I thought `fail` would _also_ cause the test to fail. If it didn't do that, that's… not useful.

File: bats/tests/helpers/paths.bash
Comment (mook-as) [inline bats/tests/helpers/paths.bash:34]: To clarify what I meant:
```suggestion
        if [ -d "$system" ]; then
            RD_LOCATION=system
        elif [ -d "$user" ]; then
            RD_LOCATION=user
        elif [ -d "$dist" ]; then
            RD_LOCATION=dist
        elif inside_repo_clone; then
            RD_LOCATION=npm
        else
            (
                echo "Couldn't locate Rancher Desktop in"
                echo "- \"$system\""
                echo "- \"$user\""
                echo "and 'npm run dev' is unavailable outside repo clone"
            ) >&3
            exit 1
        fi
```

File: bats/tests/helpers/vm.bash
Comment (mook-as) [reply bats/tests/helpers/vm.bash:35]: Yeah, definitely not as part of this PR.

**Comment by ericpromislow** [comment]
I get this breakage on macOS after running `bats bats/tests/k8s/wordpress.bats bats/tests/registry/creds.bats` in both `main` and this branch with `~/.docker/config.json` set to
```
{
  "credsStore": "none"
}
```

Breakage:
```
/Users/ericp/workspace/rancher/desktop/bats/tests/k8s/wordpress.bats
 ✓ factory reset
 ✓ add helm repo
 ✓ start rancher desktop
 ✓ deploy wordpress
 ✓ verify wordpress was deployed
/Users/ericp/workspace/rancher/desktop/bats/tests/registry/creds.bats
 ✗ factory reset
   (from function `factory_reset' in file bats/tests/helpers/vm.bash, line 17,
    in test file bats/tests/registry/creds.bats, line 82)
     `factory_reset' failed
   kill -9 90658
   kill -9 90668
   kill -9 90672
   kill -9 90853
   91287
   91293
   91303
   91304
   91307
   91311
   91315
   92382
 ✓ start container engine
 ✓ wait for container engine
 ✗ verify credential is set correctly
   (from function `assert_output' in file bats/bats-assert/src/assert_output.bash, line 194,
    from function `verify_default_credStore' in file bats/tests/registry/creds.bats, line 107,
    in test file bats/tests/registry/creds.bats, line 100)
     `verify_default_credStore' failed

   -- output differs --
   expected : osxkeychain
   actual   : none
   --

```


If I run `bats bats/tests/registry/creds.bats` in isolation I don't get the error -- for some reason the `credsStore` field gets set to `osxkeychain` at the start of the run

I don't see this breakage when running in `main`.

I've run both sequences three times in both branches and this happens consistently.

**Review by ericpromislow — CHANGES_REQUESTED** [review]
See note on breakage in https://github.com/rancher-sandbox/rancher-desktop/pull/4573#issuecomment-1530522793

**Comment by jandubois** [comment]
> See note on breakage in [#4573 (comment)](https://github.com/rancher-sandbox/rancher-desktop/pull/4573#issuecomment-1530522793)

Please provide details on how to reproduce breakage; it works for me.

Also, if this is using `RD_LOCATION=npm`, then it should be fixed in a follow-up PR; this PR is about the structural changes needed so you can specify `RD_LOCATION`.

File: bats/tests/k8s/wordpress.bats
Comment (mook-as) [inline bats/tests/k8s/wordpress.bats:5]: This doesn't skip the test, it fails it instead.
```
 ✗ setup_file failed
   The bitnami wordpress image is not available for arm64 architecture. Skipping...
   (from function `teardown_file' in test file tests/k8s/../helpers/load.bash, line 38)
     `teardown_file() {' failed with status 0
```

**Review by mook-as — CHANGES_REQUESTED** [review]

File: bats/tests/k8s/wordpress.bats
Comment (jandubois) [reply bats/tests/k8s/wordpress.bats:5]: Indeed. I'm going to revert this. Unfortunately there doesn't seem to be a way to skip the whole file, but I guess that is fine:

```console
$ ./bats-core/bin/bats tests/k8s/wordpress.bats
wordpress.bats
 - factory reset (skipped: The bitnami wordpress image is not available for arm64 architecture. Skipping...)
 - add helm repo (skipped: The bitnami wordpress image is not available for arm64 architecture. Skipping...)
 - start rancher desktop (skipped: The bitnami wordpress image is not available for arm64 architecture. Skipping...)
 - deploy wordpress (skipped: The bitnami wordpress image is not available for arm64 architecture. Skipping...)
 - verify wordpress was deployed (skipped: The bitnami wordpress image is not available for arm64 architecture. Skipping...)

5 tests, 0 failures, 5 skipped
```

**Review by ericpromislow — APPROVED** [review]
Because the breakage I saw on this PR is also visible in `main`, there's no reason to hold this PR back.

**Review by mook-as — DISMISSED** [review]

File: bats/tests/helpers/vm.bash
Comment (jandubois) [reply bats/tests/helpers/vm.bash]: Yes, it should. I've pushed another update.

**Review by mook-as — APPROVED** [review]

**Pull request merged**
